File: src/EnergyEfficient.java
// Denis Rozhnov
// SBRT Software Team: Autonavigation

// This class will be used to create and test the program that allows us to find energy-efficient paths on knnown uneven terrains
// This is a precursory algorithm. It is used for testing purposes. Does not send actual commands to the motors.

// Further improvements:
//      Send commands to motors
//      Ensure that the robot passes near certain checkpoints
//      Use input data from sensors and computer vision
//      Represent obstacles as very steep inclines

import java.util.LinkedList;
import java.util.Queue;

public class EnergyEfficient {
    /**
     * Main method for testing the algorithm
     * @param args
     */
    public static void main(String[] args){
        MapLocation start = new MapLocation(0, 0);
        MapLocation goal = new MapLocation(10,10);
        System.out.println(PATH_FINDER(start, goal));
    }

    /**
     * This is the algorithm that will compute the path between the current location and the goal location
     * @param start
     * The location at which the robot begins its motion
     * @param goal
     * The location to which the robot travels
     * @return
     * A string representing the sequecnce of coordinate points that the robot passes through.
     */
    public static String PATH_FINDER(MapLocation start, MapLocation goal){
        Queue<MapLocation> neighbors = new LinkedList<>();              // Declares a queue corresponding to neighboring MapLocations
        neighbors.add(start);                                   // Add the initial MapLocation to the queue
        double f_cost = CALCULATE_H_COST(start, goal);          // f_cost stores the computed cost of the path passing through the starting point
        double g_cost = 0;                                      // Stores the cost of traveling to a neighboring MapLocation
        MapLocation current = start;                            // Set the current location to start

        while(!neighbors.peek().equals(null)){                         // While the neighbors queue is not empty (empty only when reaches goal)
            neighbors.remove();
            MapLocation N = new MapLocation(current.getX(), current.getY()+1);
            MapLocation E = new MapLocation(current.getX()+1, current.getY());
            MapLocation S = new MapLocation(current.getX(), current.getY()-1);
            MapLocation W = new MapLocation(current.getX()-1, current.getY());              // These are all of the 8 neighbors by compass direction
            MapLocation NE = new MapLocation(current.getX()+1, current.getY()+1);
            MapLocation NW = new MapLocation(current.getX()-1, current.getY()+1);
            MapLocation SE = new MapLocation(current.getX()+1, current.getY()-1);
            MapLocation SW = new MapLocation(current.getX()-1, current.getY()-1);


            neighbors.add(N);
            neighbors.add(E);
            neighbors.add(S);
            neighbors.add(W);                               // Add neighbors to the neighbors queue
            neighbors.add(NE);
            neighbors.add(NW);
            neighbors.add(SE);
            neighbors.add(SW);

            f_cost = CALCULATE_H_COST(neighbors.peek(), goal);                // Estimates the cost of traveling to the goal from the current cell

            for(int i = 0; i < 8; i++){                                 // For each neighbor
                if(CALCULATE_H_COST(neighbors.peek(), goal) <= f_cost) {       // if the cost of traveling from that neighbor to the goal is lower than the current cost of traveling to the goal,
                    f_cost = CALCULATE_H_COST(neighbors.peek(), goal);        // reset the current cost,
                    current = neighbors.peek();                         // set current to the cheapest neighbor
                }
                if(f_cost == Double.POSITIVE_INFINITY)              // If the cost is infinite
                    return "Path not found.";

                else if(current.equals(goal))                       // If the goal is reached
                    return CONSTRUCT_PATH(goal);                    // Run a recursive method to establish the path

                double g_cost_temp = g_cost + CALCULATE_COST(current, neighbors.peek());        // Call to another method
                double f_cost_temp = g_cost_temp + CALCULATE_H_COST(neighbors.peek(), goal);    // Call to another method
                if(f_cost_temp < f_cost){
                    try{
                        neighbors.peek().setPrevious(current);
                    }
                    catch(CloneNotSupportedException cnse){
                        System.out.println("The MapLocation object cannot be cloned.");
                    }
                    g_cost = g_cost_temp;                   // Update cost values
                    f_cost = f_cost_temp;
                }

                neighbors.remove();                                     // Remove each neighbor from the neighbors queue when processed
            }                                       // Exit for-each-neighbor loop

            if(!current.equals(goal))                   // If the goal has not been reached,
                ((LinkedList<MapLocation>) neighbors).add(current);         // Add current to the linked list so as not to violate the while loop condition
        }                                          // Exit while-not-reached-goal loop

        return "Path not found.";
    }

    /**
     * Recursive method
     * @param current
     * The current MapLocation
     * @return
     * String describing the path of the rover by coordinates
     */
    public static String CONSTRUCT_PATH(MapLocation current){
        if (!current.getPrevious().equals(null)){                       // If current has a previous node (not the initial location)
            String p = CONSTRUCT_PATH(current.getPrevious());           // recursively call on that previous node
            return p + ", " + current.toString();                       // Return the recursive String
        }
        else
            return current.toString();                                  // Stopping condition when initial location is reached
    }

    /**
     * @param i
     * A MapLocation
     * @param j
     * Another MapLocation
     * @return
     * The cost of traveling from i to j
     */
    public static double CALCULATE_COST(MapLocation i, MapLocation j){
        double MASS = 50.0;                             // Approximate mass of rover
        double K_FRICTION = 0.2;                        // Approximates coefficient of kinetic friction
        double S_FRICTION = 0.3;                        // Approximates coefficient of static friction
        double P_MAX = 10;                              // Approximates the maximum available motion power of the rover
        double V = 0.5;                                 // Approximates the velocity of the robot (assumed to be constant in this model)
        double G = 9.81;                                // Stores the gravitational constant

        double z = test_function(j.getX(), j.getY()) - test_function(i.getX(), i.getY());       // Simulated difference in height

        // Calculate the angle of inclination
        double ang_inclination = Math.atan(z/Math.sqrt(Math.pow(j.getX() - i.getX(),2) + Math.pow(j.getY() - i.getY(),2)));

        // Calculate the critical impermissible angle for uphill traversal
        double critical_angle = Math.min(Math.asin(P_MAX/(V*MASS*G*Math.sqrt(Math.pow(K_FRICTION, 2)+1))) - Math.atan(K_FRICTION) , Math.atan(S_FRICTION - K_FRICTION));

        // Calculate the critical breaking angle for downhill traversal
        double critical_breaking = -Math.atan(K_FRICTION);

        if(ang_inclination > critical_angle)
            return Double.POSITIVE_INFINITY;
        else if(ang_inclination <= critical_angle && ang_inclination > critical_breaking){
            double energy_cost = MASS*G*Math.sqrt(Math.pow(j.getX() - i.getX(), 2) + Math.pow(j.getY() - i.getY(), 2) + Math.pow(z, 2))*(K_FRICTION*Math.cos(ang_inclination) + Math.sin(ang_inclination));
            return energy_cost;
        }
        else
            return 0;
    }

    /**
     * @param now
     * Current MapLocation
     * @param goal
     * Goal MapLocation
     * @return
     * The approximate energy cost of traveling from the current MapLocation to the goal
     */
    public static double CALCULATE_H_COST(MapLocation now, MapLocation goal){
        double MASS = 50.0;                             // Approximate mass of rover
        double G = 9.81;                                // Stores the gravitational constant
        double K_FRICTION = 0.2;                        // Approximates coefficient of kinetic friction
        double S_FRICTION = 0.3;                        // Approximates coefficient of static friction
        double P_MAX = 10;                              // Approximates the maximum available motion power of the rover
        double V = 0.5;                                 // Approximates the velocity of the robot (assumed to be constant in this model)


        // Calculate the critical impermissible angle for uphill traversal
        double critical_angle = Math.min(Math.asin(P_MAX/(V*MASS*G*Math.sqrt(Math.pow(K_FRICTION, 2)))) - Math.atan(K_FRICTION) , Math.atan(S_FRICTION - K_FRICTION));


        double energy = CALCULATE_COST(now, goal);
        if(energy != Double.POSITIVE_INFINITY)
            return energy;
        else{
            energy = MASS*G*(K_FRICTION*Math.cos(critical_angle) + Math.sin(critical_angle))*(test_function(goal.getX(), goal.getY()) - test_function(now.getX(), now.getY()))/Math.sin(critical_angle);
            return energy;
        }
    }

    /**
     * Function that is used to simulate the surface for test of the algorithim
     * @param x
     * x-coordinate of a MapLocation
     * @param y
     * y-coordinate of a MapLocation
     * @return
     * The elevation of the surface at a MapLocation with coordinates (x, y)
     */
    public static double test_function(int x, int y){
        return (double)x+y;
    }
}

File: Energy Efficient Path 1/EnergyEfficient.py
''' Denis Rozhnov
    SBRT Software Team: Autonavigation

 This class will be used to create and test the program that allows us to find energy-efficient paths on knnown uneven terrains
 This is a precursory algorithm. It is used for testing purposes. Does not send actual commands to the motors.
 The focus of this technique is on energy efficiency, which means that it may attempt to traverse certain obstacles
 instead of avoiding them

 This is an A* algorithm
 No CLOSED set to allow for returning to previous location

 Further improvements:
      Send commands to motors
      Ensure that the robot passes near certain checkpoints
      Use input data from sensors and computer vision
      Represent obstacles as very steep inclines

'''

import queue
from MapLocation import *
import numpy as np


# Function for finding/exploring the path
def PATH_FINDER(ns, ng):

    cost_from_start = 0                     # Total cost from start

    OPEN = queue.Queue()                    # Queue of locations currently under consideration (neighbors of interest, determined later)
    OPEN.put(ns)                            # Put the starting position into the queue

    expected_cost = CALCULATE_FROM_COST(ns, ng)             # Expected cost for starting point to goal

    current = ns                                            # Preset current to the starting position

    while not OPEN.empty():

        # Calculate the cost from the first neighbor to the goal
        if (current is not ns) or (current is ns and len(OPEN.queue) > 1):                           # If current position is not starting position
            expected_cost = EXPECTED_COST(current, OPEN.queue[0], ng, cost_from_start)          # Change calculation of expected cost

            for i in OPEN.queue:                                            # By looping through the queue
                if EXPECTED_COST(current, i, ng, cost_from_start) <= expected_cost:
                    current = i                                             # identify the neighbor with the lowest expected cost
                                                                        # set current to it
                    expected_cost = EXPECTED_COST(current, i, ng, cost_from_start)           # Update expected cost

        if expected_cost == np.inf:                 # Error message if the minimal expected cost from a neighbor to the goal is infinite
            return "Path not found."

        elif current == ng:
            #return(CONSTRUCT_PATH(current.previous, ng))                # Construct the path once the goal is reached
            return CONSTRUCT_PATH(ng)

        OPEN.get(current)                               # Remove the current location from the queue


        # Define the neighbors of the current location
        neighbors =[None]*8                       # List of neighbors

        neighbors[0] = MapLocation(current.get_x(), current.get_y()+1)
        neighbors[1] = MapLocation(current.get_x()+1, current.get_y())
        neighbors[2] = MapLocation(current.get_x(), current.get_y()-1)
        neighbors[3] = MapLocation(current.get_x()-1, current.get_y())              # Creating each neighbor
        neighbors[4] = MapLocation(current.get_x()+1, current.get_y()+1)
        neighbors[5] = MapLocation(current.get_x()-1, current.get_y()+1)
        neighbors[6] = MapLocation(current.get_x()+1, current.get_y()-1)
        neighbors[7] = MapLocation(current.get_x()-1, current.get_y()-1)

        for n in neighbors:
            cost_from_start_temp =  cost_from_start + COST_TO_NEIGHBOR(current, n)      # Projected cost from start to neighbor
            total_energy_cost_temp = cost_from_start_temp + CALCULATE_FROM_COST(n, ng)  # Projected total energy if path from current
            if EXPECTED_COST(current, n, ng, cost_from_start) is None or total_energy_cost_temp < expected_cost:
                n.set_previous(current)                         # For each neighbor of interest, set its previous to current location
                cost_from_start = cost_from_start_temp          # Update cost from start
                expected_cost = total_energy_cost_temp          # Update expected cost
                if n not in OPEN.queue:
                    OPEN.put(n)                                 # Add neighbor of interest to OPEN if it is not already there

    return "Path not found."



# Function for recursively constructing the path at the end of exploration
#def CONSTRUCT_PATH(previous, goal):
 #   if previous.previous is not None:
  #      p = CONSTRUCT_PATH(previous.previous, previous)             # If previous is not the starting location
   #     return p + ", " + previous.to_string()                      # recursively return path followed by previous
    #else:                                                           # If previous is the starting location
     #   return previous.to_string() + ", "                          # return it - this is the stopping condition


def CONSTRUCT_PATH(goal):
    path = goal.to_string()
    while goal.previous is not None:
        goal = goal.get_previous()
        path = goal.to_string() + ", " + path

    return path


# Cost of traveling to a neighbor
def COST_TO_NEIGHBOR(current, neighbor):

    MASS = 50                               # Mass of the rover
    K_FRICTION = 0.2                        # Coefficient of kinetic friction
    S_FRICTION = 0.3                        # Coefficient of static friction
    P_MAX = 10                              # Maximum available motion power
    V = 2                                   # Velocity of the rover (assumed to be constant in this model)
    G = 9.81                                # Gravitational constant

    # Compute the difference in elevation of current location and its neighbor
    z = test_function(neighbor.get_x(), neighbor.get_y()) - test_function(current.get_x(), current.get_y())

    # Compute the angle of inclination
    ang_inclination = np.arctan(z/np.sqrt((neighbor.get_x() - current.get_x())**2 + (neighbor.get_y() - current.get_y())**2))

    # Compute the critical impermissible angle for uphill traversal
    critical_up = min(np.arcsin(P_MAX/(V*MASS*G*np.sqrt(K_FRICTION**2 + 1))) - np.arctan(K_FRICTION), np.arctan(S_FRICTION - K_FRICTION))

    # Compute the critical breaking angle for downhill traversal
    critical_down = -np.arctan(K_FRICTION)

    if ang_inclination > critical_up:                       # If rover can't go uphill to reach neighbor
        return np.inf                                       # return infinity

    elif ang_inclination <= critical_up and ang_inclination > critical_down:            # Computation when angle of inclination is between the two critical angles
        cost = MASS*G*np.sqrt((neighbor.get_x() - current.get_x())**2 + (neighbor.get_y() - current.get_y())**2 + z**2) + z**2*(K_FRICTION*np.cos(ang_inclination) + np.sin(ang_inclination))
        return cost

    else:                           # Angle is less than the critical angle down, so no energy spending is required for motion
        return 0


def CALCULATE_FROM_COST(current, goal):

    MASS = 50  # Mass of the rover
    K_FRICTION = 0.2  # Coefficient of kinetic friction
    S_FRICTION = 0.3  # Coefficient of static friction
    P_MAX = 10  # Maximum available motion power
    V = 2  # Velocity of the rover (assumed to be constant in this model)
    G = 9.81  # Gravitational constant

    # Compute the critical impermissible angle for uphill traversal
    critical_up = min(np.arcsin(P_MAX / (V * MASS * G * np.sqrt(K_FRICTION ** 2 + 1))) - np.arctan(K_FRICTION),
                         np.arctan(S_FRICTION - K_FRICTION))

    expected_cost_from = COST_TO_NEIGHBOR(current, goal)                # Expected energy cost directly from current to goal

    if expected_cost_from != np.inf:
        return expected_cost_from

    else:                           # If the cost is infinite, use alternative formula to allow for zigzag traversal of inclines
        expected_cost_from = MASS*G*(K_FRICTION*np.cos(critical_up) + np.sin(critical_up))*(test_function(goal.get_x(), goal.get_y()) - test_function(current.get_x(), current.get_y()))/np.sin(critical_up)
        return  expected_cost_from


# Function to compute the expected cost of passing through the neighbor
def EXPECTED_COST(current, neighbor, goal, cost_from_start):
    return CALCULATE_FROM_COST(neighbor, goal) + COST_TO_NEIGHBOR(current, neighbor) + cost_from_start

# Function for testing the algorithm - simulation of terrain
def test_function(x, y):
    return 0



# This is the test run of the algorithm

start = MapLocation(0, 0)                   # Starting location
goal = MapLocation(10, 10)                  # Goal location

print(PATH_FINDER(start, goal))                    # Should be a recursive call of PATH_FINDER
